agent/task.go

```diff
@@ -201,6 +201,17 @@ func (tm *taskManager) run(ctx context.Context) {
 			}
 		case <-shutdown:
 			if cancel != nil {
+				// Wait for operation to finish if the task
+				// is in preparing state. This avoid the race
+				// between Prepare and Remove operations,
+				// which could cause orphan containers.
+				if tm.task.Status.State == api.TaskStatePreparing {
```

Comment:
This method should *never* be looking at task state to make decisions about what to do. State is only inspected or mutated in `exec.Do`.